### PR TITLE
build_elixir: use guards in otp version match

### PR DESCRIPTION
**Pull request by nirev**
This is an attempt to simplify the Erlang version match in the builder job.
There was a lot of repetition and a bit too hard to read IMO, so this merges some lines by using guards to specify `patch` versions when OTP dependencies change mid-major version.

### CODE REVIEW

**Comment by ericmj** [comment]
Looks good, the only problem I see is the release candidates since we won't know about them until they are out. So we would have to go back every time and fix it up like you have now.

BTW, did you try using `Version.compare/2`?

**Comment by nirev** [comment]
@ericmj I started doing it with `Version.compare/2` and I thought it was harder to understand. I can open a different PR and you pick what you think it's best.

Take this example:

```elixir
case ref do
      "v0" <> _ -> ["17.3"]
      "v1.0.0-" <> _ -> ["17.3"]
      "v1.0.0" -> ["17.3"]
      "v1.0.1" -> ["17.3"]
      "v1.0.2" -> ["17.3"]
      "v1.0.3" -> ["17.3"]
      "v1.0.4" -> ["17.5"]
      "v1.0.5" -> ["17.5"]
      "v1.0" <> _ -> ["17.3", "18.3"]
      "v1.1." <> _ -> ["17.5", "18.3"]
end
```

this turns into:
```elixir
cond do
  Version.compare(version, "1.0.4") == :lt -> 
    ["17.3"]
    
  Version.compare(version, "1.0.6") == :lt -> 
    ["17.5"]
    
  Version.compare(version, "1.1") == :lt -> 
    ["17.3", "18.3"]

  Version.compare(version, "1.2") == :lt -> 
    ["17.5", "18.3"]
end
```

I personally think it's harder to humanly parse.
When searching for "what's the mapping for version 1.1?" my mind goes straight to the line with `1.1`, but the otp version is actually the next one.

File: lib/bob/job/build_elixir.ex [hunk 1]
```diff
@@ -12,43 +12,22 @@ defmodule Bob.Job.BuildElixir do
...
       "v0" <> _ -> ["17.3"]
       "v1.0.0-" <> _ -> ["17.3"]
-      "v1.0.0" -> ["17.3"]
-      "v1.0.1" -> ["17.3"]
-      "v1.0.2" -> ["17.3"]
-      "v1.0.3" -> ["17.3"]
+      "v1.0." <> patch when patch in ~w(0 1 2 3) -> ["17.3"]
       "v1.0.4" -> ["17.5"]
       "v1.0.5" -> ["17.5"]
       "v1.0" <> _ -> ["17.3", "18.3"]
```
Comment (nirev) [inline lib/bob/job/build_elixir.ex:18]: Also could it be that the `v1.0x` mapping is wrong here?
The above lines tell that:

versions from `1.0.0-rc-x` to `1.0.3` work with erlang `17.3`
versions from `1.0.4` to `1.0.5` work with erlang `17.5`
versions from `1.0.6` up to latest `1.0.x` work with `17.3` and `18.3`

except that looking at https://github.com/elixir-lang/elixir/tags?after=v1.1.0 there doesn't seem to be any version after `1.0.5` and that is explicitly mapping to `17.5`

is this line irrelevant?
```elixir
      "v1.0" <> _ -> ["17.3", "18.3"]
```

or it is for mapping the `v1.0` tag?
in this case, why is it `["17.3", "18.3"]` if none of the 1.0.x versions map to `18.3` ?

File: lib/bob/job/build_elixir.ex [hunk 2]
```diff
@@ -57,10 +36,7 @@ defmodule Bob.Job.BuildElixir do
       "v1.8" -> ["20.3", "21.3", "22.3"]
```
Comment (nirev) [inline lib/bob/job/build_elixir.ex:36]: I'm confused with this. Where is this ref coming from? I thought it was from `elixir-lang/elixir` tags, but there is not `v1.8` tag there, only full versions with `vMAJOR.MINOR.PATCH`

Comment (ericmj) [reply lib/bob/job/build_elixir.ex:36]: It is the maintenance branch where we may backport fixes while the master branch will track the development for v1.9+.

File: lib/bob/job/build_elixir.ex (hunk 1 above)
Comment (ericmj) [reply lib/bob/job/build_elixir.ex:18]: This seems to be a mistake yes, it should either be `["17.5", "18.3"]` or `["17.5"]`. Probably safest to go with the latter.

**Comment by ericmj** [comment]
Ping. Do you have any questions I missed or anything else blocking this?

**Comment by nirev** [comment]
@ericmj Hey, sorry for the delay. I added #57 as another option using `Version.compare/2`
I  think it looks  cleaner, but requires hack.

WDYT?

**Comment by ericmj** [comment]
Closing in favor of the new PR.

**Pull request closed without merging**